### ISSUE TITLE
udp.py doRead() occasionally spews socket errors on Windows

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3396]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| p1mrx reported|
|-|-|
|Trac ID|trac#3396|
|Type|defect|
|Created|2008-08-25 02:17:59Z|
|Branch|https://github.com/twisted/twisted/tree/udp-doread-win-3396-2|

Users of my program have reported 2 cases in which Twisted spit out a socket error on Windows.  As far as I know, there's no way for a Twisted-based program to catch these things.  I don't know how to reproduce these cases, but they do happen in the wild.

Case 1:

```
Traceback (most recent call last):
  File "twisted\python\log.pyo", line 51, in callWithLogger
    
  File "twisted\python\log.pyo", line 36, in callWithContext
    
  File "twisted\python\context.pyo", line 59, in callWithContext
    
  File "twisted\python\context.pyo", line 37, in callWithContext
    
--- <exception caught here> ---
  File "twisted\internet\selectreactor.pyo", line 146, in _doReadOrWrite
    
  File "twisted\internet\udp.pyo", line 113, in doRead
    
socket.error: (10052, 'Network dropped connection on reset')
```

Case 2:

```
  Traceback (most recent call last):
    File "twisted\python\log.pyo", line 51, in callWithLogger
      
    File "twisted\python\log.pyo", line 36, in callWithContext
      
    File "twisted\python\context.pyo", line 59, in callWithContext
      
    File "twisted\python\context.pyo", line 37, in callWithContext
      
  ---  ---
    File "twisted\internet\selectreactor.pyo", line 146, in _doReadOrWrite
      
    File "twisted\internet\udp.pyo", line 113, in doRead
      
  socket.error: (10040, 'Message too long')
```

There may be more errors that Windows likes to invent, but I don't know if it's practical to enumerate all of them.

Attachments:

* [doread.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/843/843d5dfa5e9e7105802f203d1be8f07aa9455a88/53ab0a6457e016f66884b600101cdd29ba16b0f0.patch) (4453 bytes) - added by p1mrx on 2008-10-27 02:11:10Z - one proposed patch

<details><summary>Searchable metadata</summary>

```
trac-id__3396 3396
type__defect defect
reporter__p1mrx p1mrx
priority__normal normal
milestone__ 
branch__branches_udp_doread_win_3396_2 branches/udp-doread-win-3396-2
branch_author__khorn__itamar__itamarst khorn, itamar, itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1219630679000000 1219630679000000
changetime__1320885412000000 1320885412000000
version__None None
owner__itamar itamar
cc__exarkun cc__thijs cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

Fixing markup

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

What happens afterwards?  Do these errors permanently disable the UDP socket they occur on?  Do they go away after one or two calls?  After a few minutes?  After an external persistent condition has been modified?

**Comment by twisted-trac** [comment]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|p1mrx commented|
|-|-|

I think these errors only appear once, and the UDP socket continues to function normally afterwards.  The sockets are connectionless.

**Comment by twisted-trac** [comment]
|<img alt="p1mrx's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|p1mrx commented|
|-|-|

I've attached a patch that fixes the problem.  One thing of note is that I caused all errors to abort the doRead loop.  I'm not sure if this is necessary, but at least it's consistent.  In the best case, it avoids possible infinite loops.  In the worst case, it slightly decreases throughput.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_5"></a>|@glyph removed owner|
|-|-|

It's got a patch, the patch has got tests; looks like it's ready for review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_6"></a>|@glyph set owner to p1mrx|
|-|-|

Thanks for the patch.  In order to be applied to trunk, this needs:

  1. `@var` documentation for `_sockErrReadIgnore`.
  1. A docstring for `testRawSocketReads`.  Actually, it looks like this should be split into several tests, since it's testing several different conditions, and I'm still not 100% sure of what those conditions are based on reading this ticket :).  Each test should test one condition.  (Also, according to our coding standard, the method should be named `test_rawSocketReads`; since you're editing old code, the surrounding methods predate this aspect of the standard and are therefore poor examples.  Sorry about that.)
  1. `MockSocket` and `MockProtocol` need documentation as well; although: `MockSocket` should probably be put somewhere more central, like `twisted.test.proto_helpers`; it should be called `StringSocket` ('mock' is vague, and inaccurate in this case), and `MockProtocol` isn't necessary; you can just instantiate a `DatagramProtocol` and get the same effect.

Otherwise this looks pretty good.  Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @khorn|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_9"></a>|@khorn commented|
|-|-|

(In [28204]) Branching to 'udp-doread-win-3396'

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_10"></a>|@khorn commented|
|-|-|

(In [28205]) Apply patch

Refs: [#3396](https://github.com/twisted/twisted/issues/3396)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)<a name="note_11"></a>|@khorn commented|
|-|-|

(In [28206]) Address review comments

 * documentation for module level variables
 * rename MockSocket to StringSocket and move to twisted.test.proto_helpers
 * remove MockProtocol, use DatagramProtocol instead
 * rename testRawSocketReads to fit coding standard, split into multiple tests and add docstrings
 * added newsfile

Refs: [#3396](https://github.com/twisted/twisted/issues/3396)

**Comment by twisted-trac** [comment]
|[<img alt="khorn's avatar" src="https://avatars.githubusercontent.com/u/815352?s=50" width="50" height="50">](https://github.com/khorn)|@khorn removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @khorn<br>@exarkun set status to `new`|
|-|-|

  1 I think `test_rawSocketReadNormal` and `test_rawSocketReadImmediateError` need some assertions.
  1. These new tests all probably should be someplace else.  There's a parallel in `twisted/test/test_tcp_internals.py`.  That might also give you some ideas about what assertions can be added to the above tests.  Ideally there would also be something similar to the platform assumptions tests there, although I don't know to what extent this is possible for some of the errors being handled in this branch.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst commented|
|-|-|

I addressed both review comments and did a little cleanup. However, I did not really see how the platform assumption tests were quite relevant, so I didn't change the tests much.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/udp-doread-win-3396 for test results. To be honest, they're still running as I type this, but I want to go do something else, if they fail I'll take it out of review :).

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)|@zseil set owner to @zseil<br>@zseil set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_18"></a>|@zseil set owner to @itamarst<br>@zseil set status to `new`|
|-|-|

Thanks for picking this up, it's nice to see someone working on old Windows tickets!

Review comments:
 1. The branch should be merged with trunk, it currently fails on Python 2.7 builders.
 2. In topfile, "... should handle socket errors better ..." -> "... now handles socket errors better ..." or something in that vein.
 3. MSDN lists `WSAETIMEDOUT` as another error that can be returned by `recvfrom`. From the description it looks like it should be added to `_sockErrReadRefuse` error codes.
 4. `twisted.internet.iocpreactor.udp` would need a similar fix, but that can probably be done in a separate ticket. Let me know if I should open it or if you would prefer to include the fix here.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_19"></a>|@itamarst commented|
|-|-|

(In [33087]) Branching to 'udp-doread-win-3396-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_20"></a>|@itamarst removed owner|
|-|-|

Ready for review:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/udp-doread-win-3396-2

I'll open an IOCP-specific ticket once this is approved.

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)|@zseil set owner to @zseil<br>@zseil set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="zseil's avatar" src="https://avatars.githubusercontent.com/u/421894?s=50" width="50" height="50">](https://github.com/zseil)<a name="note_22"></a>|@zseil set owner to @itamarst<br>@zseil set status to `new`|
|-|-|

There is an unrelated change to `twisted.test.proto_helpers` in the branch, with that removed I think this is good to merge.

I opened the IOCP ticket myself ([#5376](https://github.com/twisted/twisted/issues/5376)), I hope you don't mind. The tests in that ticket and in [#5377](https://github.com/twisted/twisted/issues/5377) can reproduce the errors from the original description and are fixed by this branch.

Thanks everyone for working on this!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

  1.  The new test module probably belongs in `twisted.internet.test`.
  1. `StringUDPSocket` has a `@param` in its class docstring, where it doesn't make any sense.
  1. The parallel with `test_tcp_internals.py` is a bit incomplete.  The new tests here use a fake UDP socket, but don't verify the behavior of that fake.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_24"></a>|@itamarst commented|
|-|-|

Re 3: So I should not call it test_udp_internals and that'll address your review comment?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_25"></a>|@exarkun commented|
|-|-|

I don't know if you have to do anything to address it.  Ideally fakes would be verified, though.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_26"></a>|@itamarst commented|
|-|-|

Re 1, the module already is in `twisted.internet.test`, so I'm going to address this comment by doing nothing.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_27"></a>|@itamarst set status to `closed`|
|-|-|

(In [33098]) Merge udp-doread-win-3396-2.

Author: p1mrx, khorn, itamar
Review: glyph, exarkun, zseil
Fixes: [#3396](https://github.com/twisted/twisted/issues/3396)

Handle errors in UDP.doRead better: some errors can be ignored, others indicate connection refused.
